satpy/writers/ninjogeotiff.py

```diff
@@ -0,0 +1,214 @@
+GeoTIFF images, with required metadata encoded as a set of XML tags in the
+GDALMetadata TIFF tag.  Each of the XML tags must be prepended with 'NINJO_'.
+"""
+
+from .geotiff import GeoTIFFWriter
+
+
+class NinJoGeoTIFFWriter(GeoTIFFWriter):
+    """Writer for GeoTIFFs with NinJo tags."""
+
+    def save_dataset(self, dataset, ninjo_tags):
+        """Save dataset along with NinJo tags.
+
+        Save dataset along with NinJo tags.  Interface as for GeoTIFF, except
+        it takes an additional keyword argument ninjo_tags.  Those tags will be
+        prepended with ninjo_ and added as GDALMetaData.
+
+        Args:
+            dataset (xr.DataArray): Data array to save.
+            ninjo_tags (Mapping[str, str|numeric): tags to add
+        """
+        tags = calc_tags_from_dataset(dataset, ninjo_tags)
+        super().save_dataset(
+                dataset,
+                tags={"ninjo_" + k: v for (k, v) in tags.items()})
+
+
+class NinJoTagGenerator:
+    """Class to calculate NinJo tags from content."""
+
+    # tags that never change
+    fixed_tags = {
+        "Magic": "NINJO",
+        "HeaderVersion": 2,
+        "XMinimum": 1,
+        "YMinimum": 1,
+        "PhysicValue": "unknown"}
+
+    # tags that should be passed directly by the user
+    passed_tags = {"ChannelID", "DataSource", "DataType", "PhysicUnit",
+                   "SatelliteNameID"}
+
+    # tags that are calculated dynamically
+    dynamic_tags = {
+        "AxisIntercept": "axis_intercept",
+        "CentralMeridian": "central_meridian",
+        "ColorDepth": "color_depth",
+        "CreationDateID": "creation_date_id",
+        "DateID": "date_id",
+        "EarthRadiusLarge": "earth_radius_large",
+        "EarthRadiusSmall": "earth_radius_small",
+        "FileName": "filename",
+        "Gradient": "gradient",
+        "IsAtmosphereCorrected": "atmosphere_corrected",
+        "IsBlackLineCorrection": "black_line_corrected",
+        "IsCalibrated": "is_calibrated",
+        "IsNormalized": "is_normalized",
+        "MaxGrayValue": "max_gray_value",
+        "MeridianEast": "meridian_east",
+        "MeridianWest": "meridian_west",
+        "MinGrayValue": "min_gray_value",
+        "Projection": "projection",
+        "ReferenceLatitude1": "ref_lat_1",
+        "ReferenceLatitude2": "ref_lat_2",
+        "TransparentPixel": "transparent_pixel",
+        "XMaximum": "xmaximum",
+        "YMaximum": "ymaximum"
+        }
+
+    def __init__(self, dataset, args):
+        """Initialise tag generator."""
+        self.dataset = dataset
+        self.args = args
+        self.tag_names = (self.fixed_tags.keys() |
+                          self.passed_tags |
+                          self.dynamic_tags.keys())
+
+    def get_all_tags(self):
+        """Get a dictionary with all tags for NinJo."""
+        D = {}
+        for tag in self.tag_names:
+            value = self.get_tag(tag)
+            D[tag] = value
+        return D
```

Comment:
Possible one-liner that may help or hurt cleanliness:

```python
return {tag: self.get_tag(tag) for tag in self.tag_names}
```

But otherwise, are all of these tags required for all datasets and all projections? What happens if one of the tags can't be computed?

Reply:
That's an open question I still need to think about and clarify with the NinJo team.  I think not all tags are required, but I also think not even the NinJo team is sure about that.  Probably the writer could be more permissive.